### PR TITLE
add warning when using gradient_checkpointing with FSDP full shard

### LINKED ISSUE #30404: [FSDP] redundant additional allgather during backward when using FSDP FULL_SHARD with gradient checkpointing
### System Info

- `transformers` version: 4.39.3
- Platform: Linux-5.15.138.1-4.cm2-x86_64-with-glibc2.35
- Python version: 3.10.2
- Huggingface_hub version: 0.22.2
- Safetensors version: 0.4.2
- Accelerate version: 0.29.3
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes

### Who can help?

@pacman100 @muellerzr 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

**For reproduction script, please refer to this gist https://gist.github.com/yundai424/1b6f0fa9f23796033aaf585ae9e1d4b6**

Input alpaca data can be downloaded via https://huggingface.co/datasets/QingyiSi/Alpaca-CoT/blob/1d971b5591d198b810ccca6feaed420df9f39d28/alpaca/alpaca_data.json

If run the script as-is (which is essentially very basic finetuning, enabling fsdp and grad checkpointing via `--fsdp` and `--gradient_checkpointing` flags to `TrainingArguments`), it'll use the existing way of enabling gradient checkpointing, which is directly through `torch.utils.checkpoint`. By looking at the profiled trace view, it's obvious that there are 2 allgather ops per transformer block backward, and if tracing the call stack a bit you'll find the 1st one corresponds to the allgather for gradient checkpointing forward, and the 2nd one corresponds to the actual `pre_backward` hook for FSDP. This introduces unnecessary redundancy because the time gap between them is very short (i.e. the forward compute in gradient checkpointing) and is resulting in 2 -> 3 collectives per backprop (hence 3 -> 4 collectives per training step):

Screenshot 2024-04-22 at 1 02 45 PM

If uncomment [line 67-75](https://gist.github.com/yundai424/1b6f0fa9f23796033aaf585ae9e1d4b6#file-training-py-L67-L75) in the training script, it uses a different torch gradient checkpointing implementation `torch.distributed.algorithm._checkpoint` as also suggested by Meta's official [llama recipe](https://github.com/meta-llama/llama-recipes/blob/main/src/llama_recipes/policies/activation_checkpointing_functions.py). It's another layer wrapping on top of `torch.utils.checkpoint`. This will address the redundant allgather as described above, resulting in trace that looks like follows:

Screenshot 2024-04-22 at 1 03 21 PM

Note that the relative order of wrapping with `FSDP` module and calling `apply_activation_checkpointing` doesn't matter here. The above code example call `apply_activation_checkpointing` before wrapping with FSDP module, but it works in the other way too.

### Expected behavior

When using FSDP with gradient checkpointing, should instead use `torch.distributed.algorithm._checkpoint.apply_activation_checkpointing`

Will be happy to draft a PR.

### PR DESCRIPTION
**Pull request by yundai424**
# What does this PR do?

Add a warning when using FSDP full shard with `gradient_checkpointing` training arg to encourage users to use fsdp config's `activation_checkpointing` instead.

Fixes #30404

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@SunMarc  @muellerzr

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1794,6 +1794,13 @@ def __post_init__(self):
         elif FSDPOption.FULL_SHARD in self.fsdp and FSDPOption.SHARD_GRAD_OP in self.fsdp:
             raise ValueError("`--fsdp full_shard` is not compatible with `--fsdp shard_grad_op`.")
 
+        if FSDPOption.FULL_SHARD in self.fsdp and self.gradient_checkpointing:
```
Comment (SunMarc) [inline src/transformers/training_args.py:1797]: Is this only the case when we have FSDPOption.FULL_SHARD ?

**Review by SunMarc — APPROVED** [review]
LGTM @yundai424  ! Thanks a lot for adding this ! cc @muellerzr

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -1783,7 +1783,7 @@ def __post_init__(self):
             raise ValueError("warmup_steps must be either 0 or > 1")
 
         if isinstance(self.fsdp, bool):
-            self.fsdp = "full_shard" if self.fsdp else ""
+            self.fsdp = FSDPOption.FULL_SHARD if self.fsdp else ""
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1786]: From the logic below it looks like this should be a list 

```suggestion
            self.fsdp = [FSDPOption.FULL_SHARD] if self.fsdp else ""
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -1794,6 +1794,13 @@ def __post_init__(self):
         elif FSDPOption.FULL_SHARD in self.fsdp and FSDPOption.SHARD_GRAD_OP in self.fsdp:
             raise ValueError("`--fsdp full_shard` is not compatible with `--fsdp shard_grad_op`.")
 
+        if FSDPOption.FULL_SHARD in self.fsdp and self.gradient_checkpointing:
+            logger.warning(
+                "When using FSDP full shard, instead of using `gradient_checkpointing` in TrainingArguments, please"
+                " use `activation_checkpointing` in `fsdp_config`. The former one introduces a redundant AllGather"
+                " operation in backward pass. Reference: https://github.com/huggingface/transformers/issues/30404"
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1801]:
```suggestion
                " use `activation_checkpointing` in `fsdp_config`. The former introduces a redundant AllGather"
                " operation in backward pass. Reference: https://github.com/huggingface/transformers/issues/30404"
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

One question and one nit

File: src/transformers/training_args.py (hunk 1 above)
Comment (yundai424) [reply src/transformers/training_args.py:1797]: @SunMarc ah good point, hybrid shard also needs it. Update it now

**Review by muellerzr — APPROVED** [review]
Thanks!

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

@yundai424 Running `make fixup` should resolve the failing quality checks on the CI
